pkg/volume/csi/csi_mounter.go

```diff
@@ -0,0 +1,204 @@
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package csi
+
+import (
+	"errors"
+	"fmt"
+	"path"
+
+	"github.com/golang/glog"
+	grpctx "golang.org/x/net/context"
+	api "k8s.io/api/core/v1"
+	storage "k8s.io/api/storage/v1alpha1"
+	meta "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/kubernetes"
+	kstrings "k8s.io/kubernetes/pkg/util/strings"
+	"k8s.io/kubernetes/pkg/volume"
+)
+
+type csiMountMgr struct {
+	k8s        kubernetes.Interface
+	csiClient  csiClient
+	plugin     *csiPlugin
+	driverName string
+	volumeID   string
+	readOnly   bool
+	spec       *volume.Spec
+	pod        *api.Pod
+	podUID     types.UID
+	options    volume.VolumeOptions
+	volumeInfo map[string]string
+	volume.MetricsNil
+}
+
+// volume.Volume methods
+var _ volume.Volume = &csiMountMgr{}
+
+func (c *csiMountMgr) GetPath() string {
+	return getTargetPath(c.podUID, c.driverName, c.volumeID, c.plugin.host)
+}
+
+func getTargetPath(uid types.UID, driverName string, volID string, host volume.VolumeHost) string {
+	// driverName validated at Mounter creation
+	// sanitize (replace / with ~) in volumeID before it's appended to path:w
+	driverPath := fmt.Sprintf("%s/%s", driverName, kstrings.EscapeQualifiedNameForDisk(volID))
+	return host.GetPodVolumeDir(uid, kstrings.EscapeQualifiedNameForDisk(csiPluginName), driverPath)
+}
+
+// volume.Mounter methods
+var _ volume.Mounter = &csiMountMgr{}
+
+func (c *csiMountMgr) CanMount() error {
+	//TODO (vladimirvivien) use this method to probe controller using CSI.NodeProbe() call
+	// to ensure Node service is ready in the CSI plugin
+	return nil
+}
+
+func (c *csiMountMgr) SetUp(fsGroup *int64) error {
+	return c.SetUpAt(c.GetPath(), fsGroup)
+}
+
+func (c *csiMountMgr) SetUpAt(dir string, fsGroup *int64) error {
+	glog.V(4).Infof(log("Mounter.SetUpAt(%s)", dir))
+
+	ctx, cancel := grpctx.WithTimeout(grpctx.Background(), csiTimeout)
+	defer cancel()
+
+	csi := c.csiClient
+	pvName := c.spec.PersistentVolume.GetName()
+
+	// ensure version is supported
+	if err := csi.AssertSupportedVersion(ctx, csiVersion); err != nil {
+		glog.Errorf(log("failed to assert version: %v", err))
+		return err
+	}
+
+	// TODO (vladimirvivien) make assertion calls private inside client.
+	// assert volume attachment capability from driver,
+	// if not supported, assume vol is not previously attached and not ready for mounting
+	if err := csi.AssertVolumePublishCapability(ctx); err != nil {
```

Comment:
This check is not for `NodePublish` only for `ControllerPublish`. Having it here will break functionality, because in most cases, the plugin that kubelet is talking to will not have this capability. Please remove this check (and remove associated code if it is not being used anywhere else).

Reply:
@saad-ali I need further clarity on this because the spec is misleading saying that the CO must do this check:
 ```
This RPC is called by the CO when a workload that wants to use the specified volume
is placed (scheduled) on a node. The Plugin SHALL assume that this RPC will be executed 
on the node where the volume will be used. This RPC MAY be called by the CO multiple times 
on the same node for the same volume with possibly different target_path and/or auth 
credentials. If the corresponding Controller Plugin has PUBLISH_UNPUBLISH_VOLUME 
controller capability, the CO MUST guarantee that this RPC is called after ControllerPublishVolume 
is called for the given volume on the given node and returns a success.
```

Reply:
The key bit is: `If the corresponding Controller Plugin has PUBLISH_UNPUBLISH_VOLUME controller capability, the CO MUST guarantee that this RPC is called after ControllerPublishVolume is called for the given volume on the given node and returns a success.`

This means that if a plugin implements `ControllerPublish` (has PUBLISH_UNPUBLISH_VOLUME capability), then `ControllerPublish` must be successfully executed before `NodePublish` is called. We do this naturally with our attach/mount workflow. But regardless of if a controller implements `ControllerPublish` (has PUBLISH_UNPUBLISH_VOLUME capability), we must call `NodePublish`. So this check should be removed.